pkg/quota/v1/evaluator/core/resource_claims.go

```diff
@@ -98,11 +121,115 @@ func (p *claimEvaluator) MatchingResources(items []corev1.ResourceName) []corev1
+}
+
+func (p *claimEvaluator) addExtendedResourceQuota(resourceMap map[corev1.ResourceName]resource.Quantity, isExtendedResourceClaim bool, reqs corev1.ResourceList) {
+	extendedResourceMap := make(map[corev1.ResourceName]resource.Quantity)
+	for resourceName, quantity := range resourceMap {
+		if !strings.HasSuffix(resourceName.String(), corev1.ResourceClaimsPerClass) {
+			continue
+		}
+		deviceClassName := strings.TrimSuffix(string(resourceName), corev1.ResourceClaimsPerClass)
+		implicitExtendedResourceClaim := deviceClassToQuotaRequestResource(deviceClassName)
+		// implicitQuota is what has been accounted for via pod evaluator
+		// device request usage up to implicitQuota does not need to be accounted for anymore.
+		implicitQuota := reqs[implicitExtendedResourceClaim]
+		implicitQuantity := quantity
+		if isExtendedResourceClaim {
+			var zero resource.Quantity
+			if implicitQuota.Cmp(quantity) > 0 {
+				implicitQuantity = zero
+			} else {
+				implicitQuantity.Sub(implicitQuota)
+			}
+		}
+		if implicitQuantity.CmpInt64(0) > 0 {
+			extendedResourceMap[implicitExtendedResourceClaim] = implicitQuantity
+		}
+
+		var explicitQuota resource.Quantity
+		explicitQuantity := quantity
+		extendedResourceClaim, ok := p.extendedResourceQuota(deviceClassName)
+		if ok {
+			// explicitQuota is what has been accounted for via pod evaluator
+			// device request usage up to explicitQuota does not need to be accounted for anymore.
+			explicitQuota = reqs[extendedResourceClaim]
+			if isExtendedResourceClaim {
+				var zero resource.Quantity
+				if explicitQuota.Cmp(quantity) > 0 {
+					explicitQuantity = zero
+				} else {
+					explicitQuantity.Sub(explicitQuota)
+				}
+			}
+			if explicitQuantity.CmpInt64(0) > 0 {
+				extendedResourceMap[extendedResourceClaim] = explicitQuantity
+			}
+		}
+	}
+	for resourceName, quantity := range extendedResourceMap {
+		resourceMap[resourceName] = quantity
+	}
+}
+
+// Verify the claim's owner reference pod exists, and the pod's ExtendedResourceClaimStatus points
+// back to the claim if it's not nil.
+func (p *claimEvaluator) verifyOwner(claim *resourceapi.ResourceClaim) (corev1.ResourceList, bool) {
+	controllerRef := metav1.GetControllerOf(claim)
+	if controllerRef == nil {
+		return nil, false
+	}
+	if controllerRef.Kind != "Pod" || controllerRef.APIVersion != "v1" {
+		return nil, false
+	}
+	if p.podsGetter == nil {
+		return nil, false
+	}
+	pod, err := p.podsGetter.Pods(claim.Namespace).Get(controllerRef.Name)
+	if err != nil {
+		return nil, false
+	}
+	if controllerRef.UID != pod.UID {
+		return nil, false
+	}
+	if pod.Status.ExtendedResourceClaimStatus != nil && pod.Status.ExtendedResourceClaimStatus.ResourceClaimName != claim.Name {
+		return nil, false
+	}
+	reqs := resourcehelper.PodRequests(pod, resourcehelper.PodResourcesOptions{})
+	quotaReqs := corev1.ResourceList{}
+	for r, q := range reqs {
+		quotaReqs[corev1.ResourceName(corev1.DefaultResourceRequestsPrefix+string(r))] = q
+	}
```

Comment:
Use the same mechanism pod quota does to extract usage from the pod (`PodUsageFunc(pod, clock.RealClock{})`). That simplifies this code and also makes sure resourceclaim quota only exempts things pod quota actually counted.

(my commit at https://github.com/liggitt/kubernetes/commits/admit_quota/ does this)